keps/sig-network/20190603-EndpointSlice-API.md

```diff
@@ -413,28 +417,165 @@ On Pod Create/Update/Delete:
+into multiple EndpointSlices. This will occur if an Endpoints resource has
+multiple subsets or includes endpoints with multiple IP families (IPv4 and
+IPv6). Each mirrored EndpointSlice resource will:
+* Be tied to the corresponding Endpoints resource with an OwnerReference for
+  automatic garbage collection.
+* Have the `kubernetes.io/service-name` label set to the name of the Endpoints
+  resource for easy lookup.
+* Have the `endpointslice.kubernetes.io/managed-by` label set to
+  `endpointslicemirroring-controller.k8s.io` to ensure that the EndpointSlice
+  controller does not modify these EndpointSlices.
+* Mirror all labels from the Endpoints resource and all endpoints and ports
+  from the corresponding subset.
+
+#### Handling Endpoints Events
+The EndpointSliceMirroring controller will have similar logic to the existing
+EndpointSlice controller. There may be opportunities to share some code between
+the implementations.
+* **Create**: The controller will create equivalent EndpointSlices, grouped by
+  unique port and IP family combinations.
+* **Update**: The controller will compare the existing EndpointSlices mirrored
+  for this Endpoints resource with the desired set. Similar to the EndpointSlice
+  controller, it will group endpoints by unique port and IP family combinations
+  and attempt to minimize changes to EndpointSlices by recycling resources
+  wherever possible.
+* **Delete**: The controller will ensure that associated EndpointSlices are
+  removed. Although the OwnerReference should ensure that associated
+  EndpointSlices are automatically removed if the Endpoints resource is deleted,
+  manual cleanup will be required if only the skip label is changed.
+
+The controller will use a label filtered watch, so changing the skip label will
+naturally appear as create or delete events, even if the Endpoints resource is
+only modified.
+
+#### Controller Start Up
+The EndpointSliceMirroring controller will not be started until the Endpoints
+controller indicates that it has completed the first full sync. This will ensure
+that it does not mirror Endpoints that are already managed by the Endpoints
+controller (and therefore would also have EndpointSlices managed by the
+EndpointSlice controller).
+
+When the controller starts up, it will be responsible for cleaning up any excess
+EndpointSlices that it manages that are no longer desired. This will be
+determined by listing EndpointSlices the controller is managing with the
+`endpointslice.kubernetes.io/managed-by` label and comparing those with the
+EndpointSlices desired for any Endpoints with the
+`endpointslicemirroring.kubernetes.io/skip` label set to "true".
+
+#### Limitations
+To simplify implementation and align with the planned limitation on the size of
+Endpoints, the EndpointSliceMirroring controller will limit the number of
+endpoints mirrored to EndpointSlice resources to 1000 per Endpoints resource.
+
+#### Testing Plan
```

Comment:
should we define what's going to be unit , integration or e2e? is too much granularity?

Reply:
Good call, I've updated the KEP to specify that, with most seeming to work well as unit tests.